File: Top_Interview_Questions/Hard/10. Regular Expression Matching/solution4.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        s_length, p_length = len(s), len(p)
        dp = [[False for _ in range(p_length + 1)] for _ in range(s_length + 1)]
        dp[s_length][p_length] = True
        for i in range(s_length, -1, -1):
            for j in range(p_length - 1, -1, -1):
                if j + 1 < p_length and p[j + 1] == "*":
                    dp[i][j] = dp[i][j + 2] or (
                        i < s_length and (p[j] == s[i] or p[j] == ".") and dp[i + 1][j]
                    )
                else:
                    dp[i][j] = i < s_length and (p[j] == s[i] or p[j] == ".") and dp[i + 1][j + 1]
        return dp[0][0]

File: Top_Interview_Questions/Hard/10. Regular Expression Matching/solution4.cpp
#include <string>
#include <vector>
using namespace std;

/**
 * 參考解答: 用 dp table法，將遞迴的 memo 過程轉變成 for loop，dp[i][j] 代表
 * s[0..i) 是否 match p[0..j)
 * 時間: O(m*n), 空間: O(m*n)
 */

class Solution
{
  public:
    bool isMatch(string s, string p)
    {
        vector<vector<bool>> dp(s.length() + 1, vector<bool>(p.length() + 1, false));
        // since we bottom-up recursion, we use bottom-up for loop
        // (you can change it to top-down if you want)
        dp[s.length()][p.length()] = true; // base case
        for (int i = s.length(); i >= 0; i--)
        {
            for (int j = p.length() - 1; j >= 0; j--)
            {
                if (j + 1 < p.length() && p[j + 1] == '*')
                {
                    dp[i][j] = dp[i][j + 2] || (i < s.length() && (p[j] == s[i] || p[j] == '.') && dp[i + 1][j]);
                }
                else
                {
                    dp[i][j] = i < s.length() && (p[j] == s[i] || p[j] == '.') && dp[i + 1][j + 1];
                }
            }
        }
        return dp[0][0];
    }
};